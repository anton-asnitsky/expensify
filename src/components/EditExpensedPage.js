import React                from 'react';

const EditExpensedPage = (props) => {
    return (
        <div>
            Editing expense with id {props.match.params.id}
        </div>
    );
};

export default EditExpensedPage;