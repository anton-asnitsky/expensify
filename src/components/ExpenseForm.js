import React from 'react';
import moment from 'moment';
import { SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

 export default class ExpenseForm extends React.Component {
    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteTextChage = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        if( amount.match(/^\d*(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        this.setState({ createdAt });
    };

    onFocuseChage = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }));
    };

    render(){
        return (
            <div>
                <form>
                    <input 
                        type="text" 
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.setState.calendarFocused}
                        onFocusChange={this.onFocuseChage}
                        id="ExpanseFormDatePicker"
                        numberOfMonths={3}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                        placeholder="Add a note for your expense"
                        value={this.state.note}
                        onChange={this.onNoteTextChage}
                    >
                    </textarea>
                    <button>Add expense</button>
                </form>
            </div>
        );
    };
}