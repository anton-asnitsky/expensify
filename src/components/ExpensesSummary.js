import React                from 'react';
import { connect }          from 'react-redux';
import selectExpensesTotal  from '../selectors/expenses-total';
import numeral              from 'numeral';

export const ExpensesSummary = (props) => {
    return(
        <div>
            Viewing { props.expensesCount } expense{ props.expensesCount > 1 && 's'} tottaling { numeral(props.expensesTotal / 100).format('$0,0.00') }.
        </div>
    );
};

const mapStateToPorps = (state) => {
    return {
        expensesTotal: selectExpensesTotal(state.expenses),
        expensesCount: state.expenses.length
    };
};

export default connect(mapStateToPorps)(ExpensesSummary);