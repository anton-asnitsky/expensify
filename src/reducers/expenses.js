const expansesReduserDefaultState = [];
export default (state = expansesReduserDefaultState, action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id == action.id){
                    return {
                        ...expense,
                        ...action.updates
                    };
                } 
                return expense;
            });
        case 'REMOVE_EXPENSE':
            console.log(action.id);
            return state.filter(({ id }) => id !== action.id);
        default:
            return state;
    }
};