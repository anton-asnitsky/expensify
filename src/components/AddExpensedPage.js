import React        from 'react';
import ExpenseForm  from './ExpenseForm';

const AddExpensedPage = () => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm />
    </div>
);

export default AddExpensedPage;