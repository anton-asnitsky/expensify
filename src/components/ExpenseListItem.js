import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => {
    return (
        <li>
            <p>{description}</p>
            <p>{amount} - {createdAt}</p>
            <button
                onClick={(e) => {
                    dispatch(removeExpense({id}));
                }}
            >Remove</button>
        </li>
    );
};

export default connect()(ExpenseListItem);