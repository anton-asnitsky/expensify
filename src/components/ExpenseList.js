import React            from 'react';
import { connect }      from 'react-redux';
import ExpenseListItem  from './ExpenseListItem';
import selectExpenses   from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        <ul>{
            props.expenses.length === 0 ?(
                <li>No expenses found</li>
            ) : (
                props.expenses.map((expense) => {
                    return (<ExpenseListItem key={expense.id} {...expense} />);
                })
            )
        }
        </ul>
    </div>
);
const mapStateToPorps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToPorps)(ExpenseList);