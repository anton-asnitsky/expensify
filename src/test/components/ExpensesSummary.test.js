import React                    from 'react';
import { shallow }              from 'enzyme';
import { ExpensesSummary }      from '../../components/ExpensesSummary';
import expenses                 from '../fixtures/expenses';
import selectExpensesTotal      from '../../selectors/expenses-total';

test('should render 1 expense summary correctly', () => {
    const mockData = [expenses[0]];
    const expensesTotal = selectExpensesTotal(mockData);
    const wrapper = shallow(
        <ExpensesSummary u
            expensesTotal={expensesTotal}
            expensesCount={mockData.length}
        />
    );

    expect(wrapper).toMatchSnapshot();
});

test('should render multiple expense summary correctly', () => {
    const expensesTotal = selectExpensesTotal(expenses);
    const wrapper = shallow(
        <ExpensesSummary 
            expensesTotal={expensesTotal}
            expensesCount={expenses.length}
        />
    );

    expect(wrapper).toMatchSnapshot();
});