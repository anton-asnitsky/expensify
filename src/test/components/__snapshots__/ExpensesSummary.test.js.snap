// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1 expense summary correctly 1`] = `
<div>
  Viewing 
  1
   expense
   tottaling 
  $1.95
  .
</div>
`;

exports[`should render multiple expense summary correctly 1`] = `
<div>
  Viewing 
  3
   expense
  s
   tottaling 
  $1,141.95
  .
</div>
`;
